psa-crypto/src/types/key_derivation_operation.rs

```diff
@@ -0,0 +1,265 @@
+            Input::Key(key_id) => {
+                let handle = key_id.handle()?;
+                Status::from(unsafe {
+                    psa_crypto_sys::psa_key_derivation_input_key(op, step.into(), handle)
+                })
+                .to_result()?;
+                key_id.close_handle(handle)
+            }
+            Input::KeyAgreement {
+                private_key,
+                peer_key,
+            } => {
+                let handle = private_key.handle()?;
+                Status::from(unsafe {
+                    psa_crypto_sys::psa_key_derivation_key_agreement(
+                        op,
+                        step.into(),
+                        handle,
+                        peer_key.as_ptr(),
+                        peer_key.len(),
+                    )
+                })
+                .to_result()?;
+                private_key.close_handle(handle)
+            }
+        }
+    }
+}
+
+/// Enumeration of supported input data for different input steps
+#[derive(Debug, Clone)]
+pub enum Input<'a> {
+    /// Byte input for key derivation
+    Bytes(&'a [u8]),
+    /// Key input for key derivation
+    Key(Id),
+    /// Output of a key agreement
+    KeyAgreement {
+        /// Private key to use in key agreement
+        private_key: Id,
+        /// Public key data of peer key to use. Must be in the same format that `psa_import_key()` accepts for the public key
+        /// corresponding to the type of private key.
+        peer_key: &'a [u8],
+    },
+}
+
+#[cfg(feature = "interface")]
+impl TryFrom<KeyDerivationOperation<'_>> for psa_crypto_sys::psa_key_derivation_operation_t {
+    type Error = Error;
+
+    fn try_from(operation: KeyDerivationOperation) -> Result<Self> {
+        let mut op = psa_crypto_sys::psa_key_derivation_operation_init();
+        if let Err(error) = {
+            Status::from(unsafe {
+                psa_crypto_sys::psa_key_derivation_setup(
+                    &mut op,
+                    operation.alg.key_derivation().into(),
+                )
+            })
+            .to_result()?;
+            operation.alg.apply_inputs_to_op(&mut op)
+        } {
+            operation.abort(op)?;
+            return Err(error);
+        }
+
+        if let Some(capacity) = operation.capacity {
+            // Maybe best to add capacity to the algorithms for static check as some don't support it
+            Status::from(unsafe {
+                psa_crypto_sys::psa_key_derivation_set_capacity(&mut op, capacity)
+            })
+            .to_result()?;
+        }
+        Ok(op)
+    }
+}
+
+impl KeyDerivationOperation<'_> {
+    /// Clears operation C struct and consumes KeyDerivationOperation struct
+    #[cfg(feature = "interface")]
+    pub(crate) fn abort(
+        self,
+        mut op: psa_crypto_sys::psa_key_derivation_operation_t,
+    ) -> Result<()> {
+        Status::from(unsafe { psa_crypto_sys::psa_key_derivation_abort(&mut op) }).to_result()
+    }
```

Comment:
Why is this a method (instead of just a function on the type)? `self` isn't used anywhere, it would remove your need to `clone` the instance in some places

Reply:
So it consumes `self`, stopping the user from trying to reuse something that has been cleared. Think it might be better just to do it like `attributes`, let the user keep their un-cleared Rust `operation` struct while clearing the C version. I think it deviates from the PSA spec as the struct is supposed to be cleared (and we can't do that for reasons already mentioned) but I can't think of any major problem with it.